src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -1039,50 +1039,82 @@ private void RemoveWorseMembers<TMember>(ArrayBuilder<MemberResolutionResult<TMe
+                }
+            }
 
-                // If we already know this is worse than something else, no need to check again.
-                if (!c1result.Result.IsValid || worse[c1Idx] == worseThanSomething)
+            for (int index = currentBestIndex + 1; index < results.Count; index++)
+            {
+                if (!results[index].Result.IsValid)
                 {
                     continue;
                 }
 
-                bool c1IsBetterThanEverything = true;
-                for (int c2Idx = 0; c2Idx < results.Count; c2Idx++)
+                var better = BetterFunctionMember(results[currentBestIndex], results[index], arguments.Arguments, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    worse[index] = worseThanSomething;
+                }
+                else if (better == BetterResult.Right)
+                {
+                    worse[currentBestIndex] = worseThanSomething;
+                    worse[index] = betterThanEverything;
+                    currentBestIndex = index;
+                }
+                else
                 {
-                    var c2result = results[c2Idx];
+                    worse[currentBestIndex] = notBetterThanEverything;
+                    worse[index] = notBetterThanEverything;
 
-                    if (!c2result.Result.IsValid)
+                    while (++currentBestIndex < results.Count)
                     {
-                        continue;
+                        if (results[currentBestIndex].Result.IsValid)
+                        {
+                            worse[currentBestIndex] = betterThanEverything;
+                            break;
+                        }
                     }
 
-                    if (c1result.Member == c2result.Member)
-                    {
-                        continue;
-                    }
+                    index = currentBestIndex;
+                }
+            }
+
+            for (int index = 0; index < results.Count; index++)
+            {
+                if (index == currentBestIndex || !results[index].Result.IsValid)
+                {
+                    continue;
+                }
 
-                    var better = BetterFunctionMember(c1result, c2result, arguments.Arguments, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
+                var better = BetterFunctionMember(results[currentBestIndex], results[index], arguments.Arguments, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    worse[index] = worseThanSomething;
+                }
+                else if (better == BetterResult.Right)
+                {
+                    worse[currentBestIndex] = worseThanSomething;
+                    break;
+                }
+                else
+                {
+                    if (worse[currentBestIndex] != worseThanSomething)
```

Comment:
> Yes, I just removed the condition, not the statement body.

I am probably missing something because I cannot see a way for us to reach this place with ```worse[currentBestIndex] != betterThanEverything```. Isn't it the whole purpose of the ```currentBestIndex``` to point to the best? Could you please explain how could that (```worse[currentBestIndex] != betterThanEverything```) happen at this point in code?